File: follow_unfollow.rb
require 'net/http'
require 'json'
require 'time'

GITHUB_TOKEN = ENV['TOKEN']
GITHUB_USERNAME = ENV['USERNAME']

HEADERS = { 'Authorization' => "token #{GITHUB_TOKEN}" }

def get_following
  following = []
  page = 1
  loop do
    url = URI("https://api.github.com/users/#{GITHUB_USERNAME}/following?per_page=100&page=#{page}")
    response = Net::HTTP.get(url)
    data = JSON.parse(response)
    break if data.empty?

    following.concat(data.map { |user| user['login'] })
    page += 1
  end
  following
end

def get_followers
  followers = []
  page = 1
  loop do
    url = URI("https://api.github.com/users/#{GITHUB_USERNAME}/followers?per_page=100&page=#{page}")
    response = Net::HTTP.get(url)
    data = JSON.parse(response)
    break if data.empty?

    followers.concat(data.map { |user| user['login'] })
    page += 1
  end
  followers
end

def follow_user(user, retries = 3)
  follow_url = URI("https://api.github.com/user/following/#{user}")
  retries.times do
    request = Net::HTTP::Put.new(follow_url, HEADERS)
    response = Net::HTTP.start(follow_url.hostname, follow_url.port, use_ssl: true) { |http| http.request(request) }
    return true if response.code.to_i == 204
    if response.code.to_i == 403
      puts 'Rate limit hit. Waiting before retrying...'
      sleep(60)
    else
      sleep(2)
    end
  end
  false
end

def follow_all_followers
  following = get_following.to_set
  followers = get_followers.to_set
  non_following = followers - following
  if non_following.empty?
    puts 'No one Left to follow back'
    return
  end

  puts "\n #{non_following.size} are left to follow back \n"
  puts '\nList of non-followers:\n'
  non_following.each_with_index do |user, index|
    if follow_user(user)
      puts "#{index + 1}. Followed #{user}."
    else
      puts "#{index + 1}. Failed to follow #{user}."
    end
    sleep(2)
  end
  puts '\nFinished processing all non-following.'
end

def unfollow_user(user, retries = 3)
  unfollow_url = URI("https://api.github.com/user/following/#{user}")
  retries.times do
    request = Net::HTTP::Delete.new(unfollow_url, HEADERS)
    response = Net::HTTP.start(unfollow_url.hostname, unfollow_url.port, use_ssl: true) { |http| http.request(request) }
    return true if response.code.to_i == 204
    if response.code.to_i == 403
      puts 'Rate limit hit. Waiting before retrying...'
      sleep(60)
    else
      sleep(2)
    end
  end
  false
end

def find_and_unfollow_non_followers
  following = get_following.to_set
  followers = get_followers.to_set
  non_followers = following - followers
  if non_followers.empty?
    puts "You don't follow anyone who doesn't follow you back."
    return
  end

  puts "\nYou follow #{non_followers.size} people who don't follow you back."
  puts '\nList of non-followers:\n'
  non_followers.each_with_index do |user, index|
    if unfollow_user(user)
      puts "#{index + 1}. Unfollowed #{user}."
    else
      puts "#{index + 1}. Failed to unfollow #{user}."
    end
    sleep(2)
  end
  puts '\nFinished processing all non-followers.'
end

follow_all_followers
find_and_unfollow_non_followers

File: follow_unfollow.rs
extern crate reqwest;
extern crate tokio;
extern crate serde_json;

use std::env;
use std::collections::HashSet;
use std::thread::sleep;
use std::time::Duration;
use reqwest::header::{HeaderMap, HeaderValue, AUTHORIZATION};
use serde_json::Value;

async fn get_following(github_username: &str, headers: &HeaderMap<HeaderValue>) -> HashSet<String> {
    let mut following = HashSet::new();
    let mut page = 1;

    loop {
        let url = format!("https://api.github.com/users/{}/following?per_page=100&page={}", github_username, page);
        let response = reqwest::get(&url).await.unwrap().json::<Value>().await.unwrap();

        if response.as_array().unwrap().is_empty() {
            break;
        }

        for user in response.as_array().unwrap() {
            following.insert(user["login"].as_str().unwrap().to_string());
        }

        page += 1;
    }

    following
}

async fn get_followers(github_username: &str, headers: &HeaderMap<HeaderValue>) -> HashSet<String> {
    let mut followers = HashSet::new();
    let mut page = 1;

    loop {
        let url = format!("https://api.github.com/users/{}/followers?per_page=100&page={}", github_username, page);
        let response = reqwest::get(&url).await.unwrap().json::<Value>().await.unwrap();

        if response.as_array().unwrap().is_empty() {
            break;
        }

        for user in response.as_array().unwrap() {
            followers.insert(user["login"].as_str().unwrap().to_string());
        }

        page += 1;
    }

    followers
}

async fn follow_user(user: &str, headers: &HeaderMap<HeaderValue>, retries: u8) -> bool {
    let follow_url = format!("https://api.github.com/user/following/{}", user);

    for _ in 0..retries {
        let client = reqwest::Client::new();
        let response = client.put(&follow_url).headers(headers.clone()).send().await.unwrap();

        if response.status().is_success() {
            return true;
        } else if response.status().as_u16() == 403 {
            println!("Rate limit hit. Waiting before retrying...");
            sleep(Duration::from_secs(60));
        } else {
            sleep(Duration::from_secs(2));
        }
    }

    false
}

async fn follow_all_followers(github_username: &str, headers: &HeaderMap<HeaderValue>) {
    let following = get_following(github_username, headers).await;
    let followers = get_followers(github_username, headers).await;
    let non_following: HashSet<_> = followers.difference(&following).collect();

    if non_following.is_empty() {
        println!("No one Left to follow back");
        return;
    }

    println!("\n{} are left to follow back\n", non_following.len());
    println!("\nList of non-followers:\n");

    for (i, user) in non_following.iter().enumerate() {
        if follow_user(user, headers, 3).await {
            println!("{}. Followed {}.", i + 1, user);
        } else {
            println!("{}. Failed to follow {}.", i + 1, user);
        }
        sleep(Duration::from_secs(2));
    }

    println!("\nFinished processing all non-following.");
}

async fn unfollow_user(user: &str, headers: &HeaderMap<HeaderValue>, retries: u8) -> bool {
    let unfollow_url = format!("https://api.github.com/user/following/{}", user);

    for _ in 0..retries {
        let client = reqwest::Client::new();
        let response = client.delete(&unfollow_url).headers(headers.clone()).send().await.unwrap();

        if response.status().is_success() {
            return true;
        } else if response.status().as_u16() == 403 {
            println!("Rate limit hit. Waiting before retrying...");
            sleep(Duration::from_secs(60));
        } else {
            sleep(Duration::from_secs(2));
        }
    }

    false
}

async fn find_and_unfollow_non_followers(github_username: &str, headers: &HeaderMap<HeaderValue>) {
    let following = get_following(github_username, headers).await;
    let followers = get_followers(github_username, headers).await;
    let non_followers: HashSet<_> = following.difference(&followers).collect();

    if non_followers.is_empty() {
        println!("You don't follow anyone who doesn't follow you back.");
        return;
    }

    println!("\nYou follow {} people who don't follow you back.", non_followers.len());
    println!("\nList of non-followers:\n");

    for (i, user) in non_followers.iter().enumerate() {
        if unfollow_user(user, headers, 3).await {
            println!("{}. Unfollowed {}.", i + 1, user);
        } else {
            println!("{}. Failed to unfollow {}.", i + 1, user);
        }
        sleep(Duration::from_secs(2));
    }

    println!("\nFinished processing all non-followers.");
}

#[tokio::main]
async fn main() {
    let github_token = env::var("TOKEN").expect("TOKEN environment variable not set");
    let github_username = env::var("USERNAME").expect("USERNAME environment variable not set");

    let mut headers = HeaderMap::new();
    headers.insert(AUTHORIZATION, HeaderValue::from_str(&format!("token {}", github_token)).unwrap());

    follow_all_followers(&github_username, &headers).await;
    find_and_unfollow_non_followers(&github_username, &headers).await;
}
